### PR TITLE
util/testkit: add concurrent test toolkit

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

we need to add test some concurrent case but need rework for every time to setup goroutine and so on...

### What is changed and how it works?

reduce duplicate work and make write tidb concurrent test easier.

and use `TestBatchInsertWithOnDuplicate` to reproduce a old bug in 2.0.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - New methods

Side effects

 - N/A

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/9645)

### CODE REVIEW

File: util/testkit/ctestkit.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+		tk.c.Assert(err, check.IsNil)
+		id := atomic.AddUint64(&connectionID, 1)
+		se.SetConnectionID(id)
+		ctx = setSession(ctx, se)
+	}
+	return ctx
+}
+
+// OpenSessionWitDB opens new session ctx if no exists one and use db.
+func (tk *CTestKit) OpenSessionWitDB(ctx context.Context, db string) context.Context {
```
Comment (jackysp) [inline util/testkit/ctestkit.go]: OpenSessionWitDB should be OpenSessionWithDB?

File: util/testkit/ctestkit.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+func (tk *CTestKit) Exec(ctx context.Context, sql string, args ...interface{}) (sqlexec.RecordSet, error) {
+	var err error
+	tk.c.Assert(getSession(ctx), check.NotNil)
+	if len(args) == 0 {
+		var rss []sqlexec.RecordSet
+		rss, err = getSession(ctx).Execute(ctx, sql)
+		if err == nil && len(rss) > 0 {
+			return rss[0], nil
+		}
+		return nil, errors.Trace(err)
```
Comment (zz-jason) [inline util/testkit/ctestkit.go]: no need to trace?

**Pull request merged**